File: 12. InsertDeleteGetRandomO1.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class InsertDeleteGetRandomO1 {

    static class RandomizedSet {
        private final List<Integer> list = new ArrayList<>();
        private final Map<Integer, Integer> idx = new HashMap<>();
        private final Random rng = new Random();

        public boolean insert(int val) {
            if (idx.containsKey(val)) return false;
            idx.put(val, list.size());
            list.add(val);
            return true;
        }

        public boolean remove(int val) {
            Integer i = idx.remove(val);
            if (i == null) return false;
            int last = list.size() - 1;
            if (i != last) {
                int lv = list.get(last);
                list.set(i, lv);
                idx.put(lv, i);
            }
            list.remove(last);
            return true;
        }

        public int getRandom() {
            return list.get(rng.nextInt(list.size()));
        }
    }

    public static void main(String[] args) {
        RandomizedSet s = new RandomizedSet();
        System.out.println(s.insert(1));    // true
        System.out.println(s.remove(2));    // false
        System.out.println(s.insert(2));    // true
        System.out.println(s.getRandom());  // 1 or 2
        System.out.println(s.remove(1));    // true
        System.out.println(s.insert(2));    // false
        System.out.println(s.getRandom());  // 2
    }
}

File: 12. InsertDeleteGetRandomO1.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cstdlib>
#include <ctime>
using namespace std;

struct RandomizedSet {
    vector<int> lst;
    unordered_map<int, int> idx;

    bool insert(int val) {
        if (idx.count(val)) return false;
        idx[val] = lst.size();
        lst.push_back(val);
        return true;
    }

    bool remove(int val) {
        auto it = idx.find(val);
        if (it == idx.end()) return false;
        int i = it->second;
        int last = lst.size() - 1;
        if (i != last) {
            int lv = lst[last];
            lst[i] = lv;
            idx[lv] = i;
        }
        lst.pop_back();
        idx.erase(it);
        return true;
    }

    int getRandom() {
        return lst[rand() % lst.size()];
    }
};

int main() {
    srand(42);
    cout << boolalpha;
    RandomizedSet s;
    cout << s.insert(1) << "\n";
    cout << s.remove(2) << "\n";
    cout << s.insert(2) << "\n";
    cout << s.getRandom() << "\n";
    cout << s.remove(1) << "\n";
    cout << s.insert(2) << "\n";
    cout << s.getRandom() << "\n";

    return 0;
}
